### PR TITLE
Loading screen

### PR DESCRIPTION
**Pull request by JenniferJa**

### CODE REVIEW

**Review by justin-phxm — DISMISSED** [review]
LGTM

**Comment by ideen1** [comment]
I'm investigating an issue with loading background not appearing on Safari.

File: client/src/contexts/AppStateContext.tsx [hunk 1]
```diff
@@ -83,6 +107,14 @@ export function AppStateContextProvider({ children }: Props) {
     }
   }, [currentAppState]);
 
+  useEffect(() => {
+    if (!currentAppState.loading) {
+      setTimeout(() => {
+        setAnimateLoading(false);
+      }, 1000);
+    }
```
Comment (brian-ngyn) [inline client/src/contexts/AppStateContext.tsx:115]: you want to clear the timeout when this component unmounts

**Comment by ideen1** [comment]
Still need DOM callback and Socket IO callback

**Review by alexwhelan12 — DISMISSED** [review]
Looks good to me!

File: client/src/components/global/Loading.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        </div>
+        <div
+          className={twMerge(
+            "animate-circle absolute flex h-96 w-96 items-center justify-center",
+            currentLoadingState === LOADINGSTAGES.PENDING
+              ? "visible"
+              : "invisible",
+          )}
+        >
+          <Image
```
Comment (burtonjong) [inline client/src/components/global/Loading.tsx:40]: The sun should be behind the car

File: client/src/components/global/LoadingWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  useEffect(() => {
+    setTimeout(() => {
+      setCurrentLoadingStage(LOADINGSTAGES.PENDING);
+    }, 1000);
+  }, []);
+
+  // Switch to drive off after app state reports loading is complete and minimum animation time is fulfilled
+  useEffect(() => {
+    if (!currentAppState.loading) {
+      setTimeout(() => {
```
Comment (justin-phxm) [inline client/src/components/global/LoadingWrapper.tsx:32]: need to clear this timeout

File: client/src/components/global/LoadingWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  const { currentAppState, confirmVisualLoadingFulfilledAndReady } =
+    useAppState();
+  const [currentLoadingState, setCurrentLoadingStage] = useState<LOADINGSTAGES>(
+    LOADINGSTAGES.DRIVE_IN,
+  );
+
+  // Switch to pending After 1 second
+  useEffect(() => {
+    setTimeout(() => {
```
Comment (justin-phxm) [inline client/src/components/global/LoadingWrapper.tsx:24]: clear timeout

**Review by justin-phxm — CHANGES_REQUESTED** [review]

**Review by justin-phxm — DISMISSED** [review]
LGTM

**Review by alexwhelan12 — APPROVED** [review]
LGTM!

**Review by burtonjong — APPROVED** [review]
wheres the weather effects??? JK LGTM

File: client/src/components/global/LoadingWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Confirm with App State that loading animation is fulfilled after car drives off screen
+  useEffect(() => {
+    if (currentLoadingState === LOADINGSTAGES.READY) {
+      const timeout = setTimeout(() => {
+        confirmVisualLoadingFulfilledAndReady();
+      }, 800);
+      return () => clearTimeout(timeout);
+    }
+  }, [currentLoadingState]);
```
Comment (brian-ngyn) [inline client/src/components/global/LoadingWrapper.tsx:48]: we shouldn't conditionally create timeouts, i don't doubt if they are actually being cleared or not

```
useEffect(() => {
    let driveInTimeout: NodeJS.Timeout;
    let pendingTimeout: NodeJS.Timeout;
    let readyTimeout: NodeJS.Timeout;
    let confirmTimeout: NodeJS.Timeout;

    // Start with DRIVE_IN stage
    driveInTimeout = setTimeout(() => {
      setCurrentLoadingStage(LOADINGSTAGES.PENDING);
    }, 1000);

    // Switch to READY stage if loading is complete
    if (!currentAppState.loading) {
      pendingTimeout = setTimeout(() => {
        setCurrentLoadingStage(LOADINGSTAGES.READY);
      }, 4000);
    }

    // Confirm visual loading fulfillment if in READY stage
    if (currentLoadingState === LOADINGSTAGES.READY) {
      confirmTimeout = setTimeout(() => {
        confirmVisualLoadingFulfilledAndReady();
      }, 800);
    }

    // Cleanup all timeouts on component unmount or state change
    return () => {
      clearTimeout(driveInTimeout);
      clearTimeout(pendingTimeout);
      clearTimeout(readyTimeout);
      clearTimeout(confirmTimeout);
    };
  }, [currentAppState.loading, currentLoadingState, confirmVisualLoadingFulfilledAndReady]);
```

File: client/src/components/global/Loading.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            currentLoadingState === LOADINGSTAGES.READY && "text-green",
+          )}
+        >
+          {
+            {
+              1: "Connecting to Helios...",
+              2: "Connecting to Helios...",
+              3: "Connected to Helios!",
+            }[currentLoadingState]
+          }
```
Comment (brian-ngyn) [inline client/src/components/global/Loading.tsx:62]: can we do it like this?

```
const LoadingMessage = {
  [LOADINGSTAGES.DRIVE_IN]: "Connecting to Helios...",
  [LOADINGSTAGES.PENDING]: "Connecting to Helios...",
  [LOADINGSTAGES.READY]: "Connected to Helios!",
};
```

and then we render the following out inside the h2
```
{LoadingMessage[currentLoadingState]}
```

File: client/src/components/global/Loading.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            priority
+            quality={50}
+            className={twMerge(
+              "z-40",
+              currentLoadingState === LOADINGSTAGES.PENDING && "animate-bump",
+              currentLoadingState === LOADINGSTAGES.READY &&
+                "animate-driveOffScreen",
+              currentLoadingState === LOADINGSTAGES.DRIVE_IN &&
+                "animate-driveInScreen",
+            )}
```
Comment (brian-ngyn) [inline client/src/components/global/Loading.tsx:25]: this looks pretty messy, we can add this:

```
const carAnimationClass = useMemo(() => {
    switch (currentLoadingState) {
      case LOADINGSTAGES.PENDING:
        return "animate-bump";
      case LOADINGSTAGES.READY:
        return "animate-driveOffScreen";
      case LOADINGSTAGES.DRIVE_IN:
        return "animate-driveInScreen";
      default:
        return "";
    }
  }, [currentLoadingState]);
```

and then change the `className={twMerge("z-40", carAnimationClass)}`

**Review by brian-ngyn — CHANGES_REQUESTED** [review]

**Comment by brian-ngyn** [comment]
I'm wondering if we can create a loading context and a HOC like this:

```
const loadingContext = createContext<>({});

export const LoadingProvider = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [loadingCount, setLoadingCount] = useState(0);

  const startLoading = () => setLoadingCount((count) => count + 1);
  const finishLoading = () => setLoadingCount((count) => Math.max(count - 1, 0));

  useEffect(() => {
    setLoading(loadingCount > 0);
  }, [loadingCount]);

  return (
    <LoadingContext.Provider value={{ loading, startLoading, finishLoading }}>
      {children}
    </LoadingContext.Provider>
  );
};

export const useLoading = () => {
  return useContext(loadingContext);
}
```

```
const withLoading = (WrappedComponent) => {
  return (props) => {
    const { startLoading, finishLoading } = useLoading();

    useEffect(() => {
      startLoading();
      return () => finishLoading();
    }, [startLoading, finishLoading]);

    return <WrappedComponent {...props} />;
  };
};
```

and then wrap our expensive components
```
const ExpensiveComponent = ({ data }) => {
  return <>it renders something expensive or does an expensive calculation</>;
};

export default withLoading(ExpensiveComponent);
```

that way we can decouple our loading state from the appstatecontext? thoughts @ideen1 @justin-phxm

**Review by justin-phxm — COMMENTED** [review]
I was thinking maybe we could simplify a lot of the logic with React Suspense and Lazy Loading. 

https://react.dev/reference/react/Suspense

https://react.dev/reference/react/lazy

**Comment by ideen1** [comment]
> I was thinking maybe we could simplify a lot of the logic with React Suspense and Lazy Loading.
> 
> https://react.dev/reference/react/Suspense
> 
> https://react.dev/reference/react/lazy

I agree but I think the issue with Suspense is that it may hide the loader before the animation is done. if we could tie in Suspense to notify us, then we can trigger our animation to finish and hide the loader

**Comment by ideen1** [comment]
> I'm wondering if we can create a loading context and a HOC like this:
> 
> ```
> const loadingContext = createContext<>({});
> 
> export const LoadingProvider = ({ children }) => {
>   const [loading, setLoading] = useState(true);
>   const [loadingCount, setLoadingCount] = useState(0);
> 
>   const startLoading = () => setLoadingCount((count) => count + 1);
>   const finishLoading = () => setLoadingCount((count) => Math.max(count - 1, 0));
> 
>   useEffect(() => {
>     setLoading(loadingCount > 0);
>   }, [loadingCount]);
> 
>   return (
>     <LoadingContext.Provider value={{ loading, startLoading, finishLoading }}>
>       {children}
>     </LoadingContext.Provider>
>   );
> };
> 
> export const useLoading = () => {
>   return useContext(loadingContext);
> }
> ```
> 
>  ```
> const withLoading = (WrappedComponent) => {
>   return (props) => {
>     const { startLoading, finishLoading } = useLoading();
> 
>     useEffect(() => {
>       startLoading();
>       return () => finishLoading();
>     }, [startLoading, finishLoading]);
> 
>     return <WrappedComponent {...props} />;
>   };
> };
> ```
> 
> and then wrap our expensive components
> 
> ```
> const ExpensiveComponent = ({ data }) => {
>   return <>it renders something expensive or does an expensive calculation</>;
> };
> 
> export default withLoading(ExpensiveComponent);
> ```
> 
> that way we can decouple our loading state from the appstatecontext? thoughts @ideen1 @justin-phxm

I like this a lot. Only thing is I don't think we should only wrap the car graphic:
1. We want the socket connection itself to be something that we wait on for loading(it's not even a visual component but rather something where we need to set that loading is complete upon socket.connected)
2. I think the entire PIS section, car graphic section could be expensive components for the purposes of the loading screen so we should wrap both of those.

**Comment by ideen1** [comment]
The loading HOC discussion will be completed in this ticket: https://uofcsolarcar.atlassian.net/browse/TEL-116

**Comment by ideen1** [comment]
LGTM

**Review by ideen1 — APPROVED** [review]
LGTM

**Review by burtonjong — APPROVED** [review]
hlmggtm,

**Review by brian-ngyn — CHANGES_REQUESTED** [review]

**Pull request merged**